SYNTHESIZER: Synthesize loop assigns before loop invariants.

qinheping opened the pull request:
Loop invariants synthesis is dependent on loop-assigns clauses. So we target non assignable-violations until all assignable-violations are fixed.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig commented on src/goto-synthesizer/cegis_verifier.cpp:
```diff
@@ -586,143 +586,159 @@ optionalt<cext> cegis_verifiert::verify()
+  // The pointer checked in the null-pointer-check violation.
+  if(violation_type == cext::violation_typet::cex_null_pointer)
+  {
+    return_cex.checked_pointer = target_violation_info.pc->condition()
+                                   .operands()[0]
+                                   .operands()[1]
+                                   .operands()[0];
```
See https://github.com/diffblue/cbmc/pull/7456/files#r1057961414: please do more sanity checking here.

qinheping replied:
Added a function to get the checked pointer from a NULL-pointer check.

tautschnig commented on src/goto-synthesizer/cegis_verifier.cpp:
```diff
@@ -53,6 +53,55 @@ static bool contains_symbol_prefix(const exprt &expr, const std::string &prefix)
+static const exprt &
+get_checked_pointer_from_null_pointer_check(const exprt &violation)
+{
+  // A NULL-pointer check is the negation of an equation between the checked
+  // pointer and a NULL pointer.
+  const exprt *equal_expr = &violation;
+
+  // Extract the equal expression in the violation.
+  while(equal_expr->id() != ID_equal)
+  {
+    INVARIANT(equal_expr->operands().size() > 0, " No equal expression found.");
+    equal_expr = &equal_expr->operands()[0];
+  }
```
Aren't you specifically expecting a `not_exprt` at the top? If so, I'd rather suggest to use `const equal_exprt &equal_expr = to_equal_expr(to_not_expr(violation).op());`

tautschnig commented on src/goto-synthesizer/cegis_verifier.cpp:
```diff
@@ -53,6 +53,55 @@ static bool contains_symbol_prefix(const exprt &expr, const std::string &prefix)
+    equal_expr = &equal_expr->operands()[0];
+  }
+
+  const exprt *lhs_pointer = &equal_expr->operands()[0];
+  const exprt *rhs_pointer = &equal_expr->operands()[1];
```
If my above code suggestion works for you then this should be `const exprt &lhs_pointer = equal_expr.lhs();` and `const exprt &rhs_pointer = equal_expr.rhs();`

tautschnig commented on src/goto-synthesizer/cegis_verifier.cpp:
```diff
@@ -53,6 +53,55 @@ static bool contains_symbol_prefix(const exprt &expr, const std::string &prefix)
+  const exprt *lhs_pointer = &equal_expr->operands()[0];
+  const exprt *rhs_pointer = &equal_expr->operands()[1];
+
+  while(lhs_pointer->type().id() != ID_pointer)
+  {
+    INVARIANT(
+      lhs_pointer->operands().size() > 0,
+      " No pointer found in lhs of NULL-pointer checks.");
+    lhs_pointer = &lhs_pointer->operands()[0];
+  }
+
+  while(rhs_pointer->type().id() != ID_pointer)
+  {
+    INVARIANT(
+      rhs_pointer->operands().size() > 0,
+      " No pointer found in lhs of NULL-pointer checks.");
+    rhs_pointer = &rhs_pointer->operands()[0];
+  }
```
Would a non-pointer-type operand really be acceptable here?

qinheping replied:
It is ```equal_exprt(pointer_object(p1), pointer_object(p2));``` where ```p1``` is a null pointer and ```p2``` is a pointer-type operand. I will make it more explicitly.

tautschnig commented on src/goto-synthesizer/cegis_verifier.cpp:
```diff
@@ -53,6 +53,55 @@ static bool contains_symbol_prefix(const exprt &expr, const std::string &prefix)
+    rhs_pointer = &rhs_pointer->operands()[0];
+  }
+
+  // NULL == ptr
+  if(*lhs_pointer == null_pointer_exprt(to_pointer_type(lhs_pointer->type())))
+  {
+    return *rhs_pointer;
+  }
+  // ptr == NULL
+  else if(
+    *rhs_pointer == null_pointer_exprt(to_pointer_type(rhs_pointer->type())))
+  {
+    return *lhs_pointer;
+  }
```
It would be great if you could make use of `is_null_pointer` for this bit of code.

tautschnig commented on src/goto-synthesizer/cegis_verifier.cpp:
```diff
@@ -586,143 +635,155 @@ optionalt<cext> cegis_verifiert::verify()
+      std::string::npos))
+  {
+    violation_type = cext::violation_typet::cex_out_of_boundary;
+  }
 
-    // If the violation is an assignable check, we synthesize assigns targets.
-    // In the case, cause loops are all loops the violation is in. We keep
-    // adding the new assigns target to the most-inner loop that does not
-    // contain the new target until the assignable violation is resolved.
+  // The violation is a null pointer check.
+  if(
+    target_violation_info.description.find("pointer NULL") != std::string::npos)
+  {
+    violation_type = cext::violation_typet::cex_null_pointer;
+  }
 
-    // For other cases, we synthesize loop invariant clauses. We synthesize
-    // invariants for one loop at a time. So we return only the first cause loop
-    // although there can be multiple ones.
+  // The violation is a loop-invariant-preservation check.
+  if(target_violation_info.description.find("preserved") != std::string::npos)
+  {
+    violation_type = cext::violation_typet::cex_not_preserved;
+  }
 
-    log.debug() << "Start to compute cause loop ids." << messaget::eom;
+  // The violation is a loop-invariant-preservation check.
+  if(
+    target_violation_info.description.find("invariant before entry") !=
+    std::string::npos)
+  {
+    violation_type = cext::violation_typet::cex_not_hold_upon_entry;
+  }
 
-    const auto &trace = checker->get_traces()[property_it.first];
+  // The violation is an assignable check.
+  if(target_violation_info.description.find("assignable") != std::string::npos)
+  {
+    violation_type = cext::violation_typet::cex_assignable;
+  }
```
I'd say this piece of code makes for a great case to be factored out into a function of its own (returning the appropriate value for `violation_type`.

tautschnig commented on src/goto-synthesizer/dump_loop_contracts.cpp:
```diff
@@ -91,16 +91,16 @@ void dump_loop_contracts(
         " assigns ";
 
-      bool in_fisrt_iter = true;
+      bool in_first_iter = true;
```
It's great that this and further typos are being cleaned up, but I'd really appreciate if those could be put in a commit of their own.

qinheping committed:
fix typos in goto-synthesizer

tautschnig commented on src/goto-synthesizer/cegis_verifier.cpp:
```diff
@@ -53,6 +53,37 @@ static bool contains_symbol_prefix(const exprt &expr, const std::string &prefix)
+  // ! (POINTER_OBJECT(NULL) == POINTER_OBJECT(ptr))
+  const equal_exprt &equal_expr = to_equal_expr(to_not_expr(violation).op());
+
+  const exprt &lhs_pointer_object = equal_expr.lhs();
+  const exprt &rhs_pointer_object = equal_expr.rhs();
+  INVARIANT(
+    lhs_pointer_object.id() == ID_pointer_object &&
+      rhs_pointer_object.id() == ID_pointer_object,
+    "Null pointer check is an equation between two pointer objects of "
+    "pointer-type operands.");
+
+  const exprt &lhs_pointer = lhs_pointer_object.operands()[0];
+  const exprt &rhs_pointer = rhs_pointer_object.operands()[0];
```
In the same vein as above: use `to_pointer_object_expr` to get the invariant check for free.

tautschnig commented on src/goto-synthesizer/cegis_verifier.cpp:
```diff
@@ -53,6 +53,37 @@ static bool contains_symbol_prefix(const exprt &expr, const std::string &prefix)
+  const exprt &lhs_pointer = lhs_pointer_object.operands()[0];
+  const exprt &rhs_pointer = rhs_pointer_object.operands()[0];
+
+  // NULL == ptr
+  if(
+    can_cast_expr<constant_exprt>(lhs_pointer) &&
+    is_null_pointer(*expr_try_dynamic_cast<constant_exprt>(lhs_pointer)))
+  {
+    return rhs_pointer;
+  }
+
```
Hmm, shouldn't you support NULL on either side of the equality? I'd expect an `else if(<same condition as above but using rhs_pointer>)`.

qinheping replied:
No, the null-pointer check is
! (POINTER_OBJECT(null) == POINTER_OBJECT(ptr))
according to ```ansi-c/goto_check_c.cpp```.

tautschnig commented on src/goto-synthesizer/cegis_verifier.cpp:
```diff
@@ -94,6 +125,45 @@ optionst cegis_verifiert::get_options()
+  {
+    violation_type = cext::violation_typet::cex_out_of_boundary;
+  }
+
+  // The violation is a null pointer check.
+  if(description.find("pointer NULL") != std::string::npos)
```
Should you be using `else if` here (and below), or is it indeed the case that multiple branches can match and you want to take the last such match?

qinheping committed:
Synthesize loop assigns before loop invariants.

Loop invariants synthesis is dependent on loop-assigns clauses. So we don't target non assignable-violations until all assignable-violations are fixed.
